src/vivarium/framework/artifact/artifact.py

```diff
@@ -0,0 +1,281 @@
+        """
+        Parameters
+        ----------
+        path
+            The path to the artifact file.
+        filter_terms
+            A set of terms suitable for usage with the ``where`` kwarg
+            for :func:`pd.read_hdf`.
+
+        """
+        self._path = Path(path)
+        self._filter_terms = filter_terms
+        self._draw_column_filter = _parse_draw_filters(filter_terms)
+        self._cache = {}
+
+        self.create_hdf_with_keyspace()
+        self._keys = Keys(self._path)
+
+    @property
+    def path(self):
+        """The path to the artifact file."""
+        return str(self._path)
+
+    @property
+    def keys(self) -> List[str]:
+        """A list of all the keys contained within the artifact."""
+        return self._keys.to_list()
+
+    @property
+    def filter_terms(self) -> List[str]:
+        """Filters that will be applied to the requested data on loads."""
+        return self._filter_terms
+
+    def create_hdf_with_keyspace(self):
+        """Creates the artifact HDF file and adds a node to track keys."""
+        if not self._path.is_file():
+            hdf.touch(self._path)
+            hdf.write(self._path, 'metadata.keyspace', ['metadata.keyspace'])
+
+    def load(self, entity_key: str) -> Any:
+        """Loads the data associated with provided entity_key.
+
+        Parameters
+        ----------
+        entity_key
+            The key associated with the expected data.
+
+        Returns
+        -------
+            The expected data. Will either be a standard Python object or a
+            :class:`pandas.DataFrame` or :class:`pandas.Series`.
+
+        Raises
+        ------
+        ArtifactException :
+            If the provided key is not in the artifact.
+
+        """
+        if entity_key not in self.keys:
+            raise ArtifactException(f"{entity_key} should be in {self.path}.")
+
+        if entity_key not in self._cache:
+            data = hdf.load(self._path, entity_key, self._filter_terms, self._draw_column_filter)
+            # FIXME: Under what conditions do we get None here.
+            assert data is not None, f"Data for {entity_key} is not available. Check your model specification."
+            self._cache[entity_key] = data
+
+        return self._cache[entity_key]
+
+    def write(self, entity_key: str, data: Any):
+        """Writes data into the artifact and binds it to the provided key.
+
+        Parameters
+        ----------
+        entity_key
+            The key associated with the provided data.
+        data
+            The data to write. Accepted formats are :class:`pandas.Series`,
+            :class:`pandas.Dataframe` or standard python types and containers.
+
+        Raises
+        ------
+        ArtifactException :
+            If the provided key already exists in the artifact.
+
+        """
+        if entity_key in self.keys:
+            raise ArtifactException(f'{entity_key} already in artifact.')
+        elif data is None:
```

Comment:
remind me again why this is here? what use case do we have for calling write with None?